### PR TITLE
Bug fix(es); Add cmdline parameter to sample a specific thread or set of threads; Code Cleanup; update build scripts

### PR DESCRIPTION
**Pull request by mojomojomojo**
## Bug Fixes

  * array out of bounds in `ProfilerThread::sample(...)`
  * ignored input parameter `ThreadPicker::AttachToProcess(...)`
  * cleaned up `assert(...)` to be able to build with Debug configuration(s)

One or more of these was resulting in bad behavior. If you select a single thread and start sampling, the app will sample a single thread once (or a few times), then begin sampling _all_ threads.

## Attach to specific thread(s)

    /thread:<tid>

    /thread:<tid1>,<tid2>,...

In addition to /mt /mbt, this option permits you to sample specific threads, by thread ID.

## Build Scripts

  * added some feedback for failures
  * added support for default install paths for Inno 6
  * defensively quote paths

### CODE REVIEW

File: src/profiler/debugger.cpp [hunk 1]
```diff
@@ -52,7 +52,7 @@ void Debugger::notifyNewThread(DWORD threadId, HANDLE threadHandle)
 bool Debugger::attach(std::function<void(NotifyData const &notification)> notifyFunc_)
 {
 	assert(!processHandle);
-	assert(!debuggerThread);
+	//assert(!debuggerThread);
```
Comment (CyberShadow) [inline src/profiler/debugger.cpp]: If there is a reason to disable the assert, please just remove it. Few things cause anxiety-filled uncertainty when trying to understand a codebase like a commented-out assert :)

File: src/profiler/profiler.cpp [hunk 2]
```diff
@@ -112,7 +112,7 @@ void applyHacks(HANDLE process_handle, CONTEXT32 &context)
 
 	// First, skip over any stub functions (a useless push/mov/pop header, followed by a jump).
 	// Move instead to the jump target.
-	if (ReadProcessMemory(process_handle, (LPCVOID)context.Eip, tmp, 16, &numRead) && numRead >= 16)
+	if (ReadProcessMemory(process_handle, reinterpret_cast<LPCVOID>((uintptr_t)context.Eip), tmp, 16, &numRead) && numRead >= 16)
```
Comment (CyberShadow) [inline src/profiler/profiler.cpp:115]: What is the purpose of this change? Does it address some compiler warning? Ideally such changes should be on their own commit and the reason for them explained in the commit message.

File: src/profiler/profiler.cpp [hunk 3]
```diff
@@ -279,7 +279,7 @@ bool Profiler::sampleTarget(SAMPLE_TYPE timeSpent, SymbolInfo *syminfo)
 			stack.addr[stack.depth++] = ip;
 		first = false;
 
-		BOOL result = dbgHelp->StackWalk64(
+		BOOL result2 = dbgHelp->StackWalk64(
```
Comment (CyberShadow) [inline src/profiler/profiler.cpp:282]: It would be better to put the previous variable in a narrower scope instead.

File: src/wxProfilerGUI/profilergui.cpp [hunk 4]
```diff
@@ -62,6 +62,7 @@ static const wxCmdLineEntryDesc g_cmdLineDesc[] =
 	{ wxCMD_LINE_SWITCH, "h", "", "Displays help on the command line parameters.",			wxCMD_LINE_VAL_NONE, wxCMD_LINE_OPTION_HELP },
 	{ wxCMD_LINE_OPTION, "r", "", "Runs an executable and profiles it.",					wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL|wxCMD_LINE_NEEDS_SEPARATOR },
 	{ wxCMD_LINE_OPTION, "a", "", "Attaches to a process (by its PID) and profiles it.",	wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL|wxCMD_LINE_NEEDS_SEPARATOR },
+	{ wxCMD_LINE_OPTION, "thread", "", "Profiles the specified thread(s) in the process, multiple threads must be in a comma-delimited list (See -a for specifying the process ID).",	wxCMD_LINE_VAL_STRING, wxCMD_LINE_PARAM_OPTIONAL|wxCMD_LINE_NEEDS_SEPARATOR },
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp:65]: Should this perhaps specify how to specify the threads (decimal thread ID?)?

File: src/wxProfilerGUI/profilergui.cpp [hunk 5]
```diff
@@ -637,6 +639,26 @@ bool ProfilerGUI::Run()
 	else if (!cmdline_attach.empty())
 	{
 		std::unique_ptr<AttachInfo> info(AttachToProcess(cmdline_attach));
+		if (!cmdline_thread_ids.empty()) {
+			std::cout << "(Orig) Profiling Threads: ";
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp]: I'm not sure of the value of leaving these debug messages here.

File: thirdparty/drmingw_build_mingw.cmd [hunk 6]
```diff
@@ -21,13 +21,15 @@ set MINGW64_FN=x86_64-6.3.0-release-win32-seh-rt_v5-rev1.7z
 set MINGW32_URL=https://sourceforge.net/projects/mingw-w64/files/Toolchains%%20targetting%%20Win32/Personal%%20Builds/mingw-builds/6.3.0/threads-win32/dwarf/%MINGW32_FN%/download
 set MINGW64_URL=https://sourceforge.net/projects/mingw-w64/files/Toolchains%%20targetting%%20Win64/Personal%%20Builds/mingw-builds/6.3.0/threads-win32/seh/%MINGW64_FN%/download
 
+REM  I downloaded these manually.
```
Comment (CyberShadow) [inline thirdparty/drmingw_build_mingw.cmd:24]: What value does this addition bring? It's not ever clear to readers who "I" is.

File: thirdparty/drmingw_build_mingw.cmd [hunk 7]
```diff
@@ -21,13 +21,15 @@ set MINGW64_FN=x86_64-6.3.0-release-win32-seh-rt_v5-rev1.7z
...
 
+REM  I downloaded these manually.
 for %%t in (32 64) do (
 	set FN=!MINGW%%t_FN!
 	set URL=!MINGW%%t_URL!
 
 	if not exist "%~dp0\mingw\!FN!" (
 		echo Downloading !FN!...
 		set DEST=%~dp0/mingw/!FN!
+                REM If the automagic download fails, download the files manually and put them in this directory.
```
Comment (CyberShadow) [inline thirdparty/drmingw_build_mingw.cmd:32]: This file uses tabs for indentation, not spaces.

File: thirdparty/drmingw_build_mingw.cmd [hunk 8]
```diff
@@ -81,6 +83,9 @@ goto :eof
 rem Download !URL! to !DEST!.
 rem Note: Positional parameters don't work here, as the percent signs in URLs are eagerly interpreted despite quoting.
 
+@echo Downloading %URL%
+@echo   %DEST%
+
```
Comment (CyberShadow) [inline thirdparty/drmingw_build_mingw.cmd:88]: This batch file uses delayed expansion, so for consistency this should use that syntax.

File: build.cmd [hunk 9]
```diff
@@ -107,10 +107,12 @@ if not %CONFIGURATION% == Release goto :eof
 rem Find InnoSetup
 
 if not defined INNOSETUP for %%a in (ISCC.exe) do if not [%%~$PATH:a] == [] set INNOSETUP="%%~$PATH:a"
-if not defined INNOSETUP if exist "%ProgramFiles%\Inno Setup 5\ISCC.exe" set INNOSETUP="%ProgramFiles%\Inno Setup 5\ISCC.exe"
-if not defined INNOSETUP if exist "%ProgramFiles(x86)%\Inno Setup 5\ISCC.exe" set INNOSETUP="%ProgramFiles(x86)%\Inno Setup 5\ISCC.exe"
-if not defined INNOSETUP if exist "%ProgramW6432%\Inno Setup 5\ISCC.exe" set INNOSETUP="%ProgramW6432%\Inno Setup 5\ISCC.exe"
-if not defined INNOSETUP if exist "%SystemDrive%\Inno Setup 5\ISCC.exe" set INNOSETUP="%SystemDrive%\Inno Setup 5\ISCC.exe"
+for %%v in (5 6) do (
```
Comment (CyberShadow) [inline build.cmd:110]: Have you tested Inno Setup 6 and found it to work perfectly (installation, installation of 32-bit on 64-bit, uninstallation, signing)?

File: src/wxProfilerGUI/profilergui.cpp [hunk 10]
```diff
@@ -728,18 +728,13 @@ bool ProfilerGUI::OnCmdLineParsed(wxCmdLineParser& parser)
...
 		for (size_t i=0; i<tids_str.GetCount(); i++) {
 			long tid;
 			if (tids_str[i].ToLong(&tid)) {
 				cmdline_thread_ids.push_back(tid);
 			}
 		}
-		std::cout << "  Threads to attach to (";
-		for (auto& tid : cmdline_thread_ids)
-			std::cout << " " << tid;
-		std::cout << " )\n";
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp]: Please squash this commit into the commit which added the debug code, so that it won't be present in the Git history.

File: src/profiler/profilerthread.cpp [hunk 11]
```diff
@@ -126,7 +126,7 @@ void ProfilerThread::sample(const SAMPLE_TYPE timeSpent)
 		}
 	}
 
-	for (ptrdiff_t n = (ptrdiff_t)count; n >= 0; --n)
+	for (ptrdiff_t n = (ptrdiff_t)count-1; n >= 0; --n)
```
Comment (CyberShadow) [inline src/profiler/profilerthread.cpp:129]: Please squash this change into the commit that added this code.

File: src/wxProfilerGUI/profilergui.cpp [hunk 12]
```diff
@@ -304,7 +304,9 @@ std::wstring ProfilerGUI::LaunchProfiler(const AttachInfo *info)
 AttachInfo::AttachInfo()
 {
 	process_handle = NULL;
+	attach_all_threads = true;
 	sym_info = NULL;
+	delay_profile = 0; // unused?
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp:309]: Yes, please delete this member instead.

File: src/wxProfilerGUI/profilergui.cpp [hunk 13]
```diff
@@ -642,25 +642,15 @@ bool ProfilerGUI::Run()
...
 					profile_threads.push_back(tid_h);
 				}
 			}
 			info->thread_handles = profile_threads;
-			std::cout << "Profiling Threads: ";
-			for (auto hwnd : info->thread_handles)
-				std::cout << " " << (uintptr_t)hwnd;
-			std::cout << "\n";
-
+			// Is there a threshold we should use to set this? (currently: attach_count < total_count)
```
Comment (CyberShadow) [inline src/wxProfilerGUI/profilergui.cpp:653]: I think attaching to select threads and also to new threads would make sense as something explicitly opt-in. In any case, please don't leave open-ended questions in comments in the source code :)

File: src/copyfiles.bat [hunk 14]
```diff
@@ -14,22 +14,22 @@ if %PLATFORM%==x64   set PLATFORM_X=x64
 
 set DBGHELPERS=dbghelp_%PLATFORM_X%
 copy /y %DBGHELPERS%\*.* %DEST%
-if errorlevel 1 exit /b 1
+if errorlevel 1 ((@ECHO Failed to copy dbghelp_*) & (exit /b 1))
```
Comment (CyberShadow) [inline src/copyfiles.bat:17]: These messages are only partially helpful. A better way to debug batch files is to remove the `@echo off` from the top, so I think these additions are not worth their weight.

**Review by CyberShadow — COMMENTED** [review]
Thanks for the submission. I'll be happy to merge this if you could please move all changes which are not strictly necessary for specifying thread IDs out of this pull request. These can be submitted and discussed as separate pull requests if you wish.

File: thirdparty/drmingw_build_mingw.cmd (hunk 6 above)
Comment (CyberShadow) [reply thirdparty/drmingw_build_mingw.cmd:24]: The downloads still work on AppVeyor, so I posit that there is a technical issue or incompatibility on your system that prevents these from being downloaded successfully.

File: src/profiler/debugger.cpp (hunk 1 above)
Comment (mojomojomojo) [reply src/profiler/debugger.cpp]: The assert won't compile, and I wasn't certain what it was referring to. (What is `debuggerThread`?) Out of caution, I chose not to remove it, thinking someone else could make a better decision.

File: src/profiler/profiler.cpp (hunk 2 above)
Comment (mojomojomojo) [reply src/profiler/profiler.cpp:115]: It creates warnings in newer compilers (e.g. VS 2017).

File: src/wxProfilerGUI/profilergui.cpp (hunk 4 above)
Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp:65]: I wanted to be able to specify multiple threads. I couldn't find anything baked-in to wxWidgets's command line framework to allow multiple values. So, it's a string that we manually split and parse.

File: src/wxProfilerGUI/profilergui.cpp (hunk 5 above)
Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp]: You're correct. I thought I'd removed these.

File: src/profiler/debugger.cpp (hunk 1 above)
Comment (CyberShadow) [reply src/profiler/debugger.cpp]: I see, it's a leftover of #88, but nobody tried to build Very Sleepy in the debug profile, so this wrongful addition went unnoticed. In this case feel free to add a commit which removes such left-overs.

File: src/copyfiles.bat (hunk 14 above)
Comment (mojomojomojo) [reply src/copyfiles.bat:17]: An alternate philosophy is to provide some meaningful information when there's a failure so that you don't have to re-run anything (potentially changing the conditions under which the failure originally occurred) to get more details.

File: src/profiler/profiler.cpp (hunk 2 above)
Comment (CyberShadow) [reply src/profiler/profiler.cpp:115]: Does `(LPCVOID)(uintptr_t)context.Eip` work just as well?

File: src/wxProfilerGUI/profilergui.cpp (hunk 4 above)
Comment (CyberShadow) [reply src/wxProfilerGUI/profilergui.cpp:65]: Right, but that's not what I mean. I mean that perhaps the help text for the switch should explain what exactly is the expected format and contents of those comma-delimited values.

File: src/wxProfilerGUI/profilergui.cpp (hunk 5 above)
Comment (CyberShadow) [reply src/wxProfilerGUI/profilergui.cpp]: I see now that they've been removed in follow-up commits. In this case please just squash the removal with the addition.

File: src/copyfiles.bat (hunk 14 above)
Comment (CyberShadow) [reply src/copyfiles.bat:17]: Good point, though the build script tries to be idempotent and only do what is needed (though it mostly delegates this part to MS Build).

Alternatively to defensive programming perhaps it would make sense to directly address the cause of the failure you were seeing?

File: src/wxProfilerGUI/profilergui.cpp (hunk 4 above)
Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp:65]: None of the other args have examples. Perhaps the format of this arg is unusual.
I don't mind doing this.

File: thirdparty/drmingw_build_mingw.cmd (hunk 6 above)
Comment (mojomojomojo) [reply thirdparty/drmingw_build_mingw.cmd:24]: Sry. Left in by accident. It was a note to myself.

File: thirdparty/drmingw_build_mingw.cmd (hunk 7 above)
Comment (mojomojomojo) [reply thirdparty/drmingw_build_mingw.cmd:32]: Bummer.

File: src/profiler/profiler.cpp (hunk 2 above)
Comment (CyberShadow) [reply src/profiler/profiler.cpp:115]: Alright. Personally I think `reinterpret_cast` makes perfect sense for types for which different kinds of conversions can be conceived (such as numeric/floating point or bytes to structs), but converting integer to pointer is fairly unambiguous, so a C cast would be OK here and have less syntactic overhead.

File: thirdparty/drmingw_build_mingw.cmd (hunk 8 above)
Comment (mojomojomojo) [reply thirdparty/drmingw_build_mingw.cmd:88]: Why use it if it's not needed? That just suggests a misunderstanding of cmd/batch.

File: build.cmd (hunk 9 above)
Comment (mojomojomojo) [reply build.cmd:110]: I've used both install/uninstall on 64-bit on 64-bit.
I can test 32-on-64.
IDK about signing.

File: thirdparty/drmingw_build_mingw.cmd (hunk 8 above)
Comment (CyberShadow) [reply thirdparty/drmingw_build_mingw.cmd:88]: It was needed somewhere, so we might as well use it everywhere for consistency. Is there a reason not to?

File: src/profiler/profilerthread.cpp (hunk 11 above)
Comment (mojomojomojo) [reply src/profiler/profilerthread.cpp:129]: I'm not sure what you mean.
The original commit (by someone else) where this bug was introduce?

File: build.cmd (hunk 9 above)
Comment (CyberShadow) [reply build.cmd:110]: Great, thank you.

File: src/profiler/profilerthread.cpp (hunk 11 above)
Comment (CyberShadow) [reply src/profiler/profilerthread.cpp:129]: Ah, my bad; I thought this code was added in an earlier commit in this pull request.

Comment (CyberShadow) [reply src/profiler/profilerthread.cpp:129]: It would be great if the commit message for this change explained the fix and mention the commit where this bug was introduced.

Comment (mojomojomojo) [reply src/profiler/profilerthread.cpp:129]: I tried blame, but apparently it looked like it has been around ever since the code was moved from C# to C++.

File: src/copyfiles.bat (hunk 14 above)
Comment (mojomojomojo) [reply src/copyfiles.bat:17]: The first problem is that I didn't know which file was failing to be copied. The script just failed quietly (`exit /b 1`). The error message from cmd/COPY didn't include details. I realize this is not the script's problem, but MS isn't likely to fix it.

File: src/wxProfilerGUI/profilergui.cpp (hunk 13 above)
Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp:653]: It was more of a question to someone more knowledgeable about the code than I. I can remove it.

File: src/copyfiles.bat (hunk 14 above)
Comment (CyberShadow) [reply src/copyfiles.bat:17]: So which file was it? Copying a file implies a precondition that the file must exist, and if execution of the script can reach this point with the file not existing, then that points to a bug somewhere earlier in the script.

File: thirdparty/drmingw_build_mingw.cmd (hunk 6 above)
Comment (mojomojomojo) [reply thirdparty/drmingw_build_mingw.cmd:24]: I've seen a number of posts complaining about the sourceforge downloads. IDK what I could be doing wrong. The script is using very simple download tools that behave identically across systems. I'd vote for moving away from SF, if possible, but I don't have the resources to do anything about this.
Either way, I should have removed this message.

File: thirdparty/drmingw_build_mingw.cmd (hunk 8 above)
Comment (mojomojomojo) [reply thirdparty/drmingw_build_mingw.cmd:88]: Using it ubiquitously seems like voodoo programming to me.
There are actually some occasions when you _don't_ want the expansion to be delayed, but this isn't one of them.
I can't think of a functional reason to avoid it other than suggesting that you don't know how the "feature" works. I only use it when it's necessary (e.g. inside parens for `FOR` or `IF`).
I prefer to be precise about the use of features, but since I didn't go back and "fix" all the other instances of it, I can at least imagine that it could be confusing to other developers.

File: src/profiler/profilerthread.cpp (hunk 11 above)
Comment (mojomojomojo) [reply src/profiler/profilerthread.cpp:129]: I must have been dreaming.

File: src/copyfiles.bat (hunk 14 above)
Comment (mojomojomojo) [reply src/copyfiles.bat:17]: It was one of the files created from the drmingw build. To anyone unacquainted with the innards of this part of the project, it seems like adding a specific note would at least enable them to look at the part of the script where the error message came from and more quickly drill down to the source of the problem.

Why is this code in the script?

    if errorlevel 1 echo copyfiles.bat: Warning: couldn't copy the C/C++ runtime DLLs

I consider it sensible, but I didn't add it. It very clearly conveys the source of the problem.

File: src/profiler/profiler.cpp (hunk 3 above)
Comment (mojomojomojo) [reply src/profiler/profiler.cpp:282]: How about if I rename them both so that neither of them has a generic name?

File: src/copyfiles.bat (hunk 14 above)
Comment (CyberShadow) [reply src/copyfiles.bat:17]: > It was one of the files created from the drmingw build.

Okay, that makes sense. It's a separate project and may change beyond our control.

> Why is this code in the script?

Because it also copies files from a location beyond our control. Unfortunately I don't remember why it's merely a warning.

File: src/wxProfilerGUI/profilergui.cpp (hunk 13 above)
Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp:653]: In `ProfilerGUI::LaunchProfiler(...)`, `attach_all_threads` is used to decide whether or not to launch a new `Debugger` thread. This change (sort of) adds a wrinkle. There could be a number of threads to profile without being _all_ of the threads.

Under what conditions do we want to launch the other debugger thread? The comments in the code don't make this entirely clear. I assume it's for performance reasons, but I don't really know when doing this is worth it.

Comment (mojomojomojo) [reply src/wxProfilerGUI/profilergui.cpp:653]: You can already select multiple threads with the GUI, and so I attempted to retain what was probably the existing behavior, WRT `attach_all_threads`. It just seemed like someone ought to make an intelligent decision about this.

File: src/profiler/profiler.cpp (hunk 3 above)
Comment (mojomojomojo) [reply src/profiler/profiler.cpp:282]: Would you rather have arbitrary code blocks to remove the variable from the function-level scope? I'm not going to fight you about what you want here, I just don't see an option that is clearly better in all respects.

Comment (CyberShadow) [reply src/profiler/profiler.cpp]: I do, and it does make sense to me to reduce visibility of declarations as much as possible, even within functions, but perhaps that's an unusual preference. Someone more pedantic about code cleanliness would probably suggest that if we are at a point where variable names clash within a function, it's time to refactor that function into several.

In any case this is such a minor issue that I'm happy with any solution. I would ask though that the fix is in its own commit (or grouped with other fixes with the same purpose) and the commit message documents why the change is done.
